[JENKINS-52822] Add CLI command to enable a plugin

jvz opened the pull request:
See [JENKINS-52822](https://issues.jenkins-ci.org/browse/JENKINS-52822).


* Added a new CLI command `enable-plugin` to enable one or more installed plugins and optionally restart Jenkins.


- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->


@tracymiranda @daniel-beck @varyvol @reviewbybees

jvz committed:
[JENKINS-52822] Add CLI command to enable a plugin

daniel-beck commented on core/src/main/java/hudson/cli/EnablePluginCommand.java:
```diff
@@ -0,0 +1,35 @@
+package hudson.cli;
```
Lacks license header.

jvz committed:
[JENKINS-52822] Add CLI command to enable a plugin

This adds the ability to enable multiple plugins at once along with their dependencies.

daniel-beck commented:
Why WIP?

jvz committed:
Add missing license headers

jvz commented:
Added transitive enabling of plugins and some better testing. WIP off.

jvz committed:
Clarify javadoc

oleg-nenashev reviewed (approved):
the code LGTM

daniel-beck commented on core/src/main/java/hudson/cli/EnablePluginCommand.java:
```diff
@@ -0,0 +1,80 @@
+        for (PluginWrapper.Dependency dependency : plugin.getDependencies()) {
+            enablePlugin(manager, dependency.shortName);
+        }
+        plugin.enable();
```
Given the magic happening with dependencies here, I'd expect every enabled plugin would appear in command output.

varyvol replied:
Yes, that sounds like a good idea.

MRamonLeon replied:
_Silly question_: could we have a cycle in the dependencies? **A** -> B -> C -> D -> **A!** If it's possible then this bucle won't finish. I don't mean if its logical, I mean if it's possible.

daniel-beck replied:
> could we have a cycle in the dependencies? 

Always a DAG. We have special support to ensure it even for detached plugins.

MRamonLeon replied:
Fine then.

daniel-beck reviewed (approved):
Not a fan of the silent operation, but looks reasonable overall.

varyvol commented on core/src/main/resources/hudson/cli/Messages.properties:
```diff
@@ -25,6 +27,8 @@ DeleteViewCommand.ShortDescription=\
 DeleteJobCommand.ShortDescription=\
  Deletes job(s).
+EnablePluginCommand.ShortDescription=\
+ Enables an installed plugin.
```
Maybe adding something like "and its dependencies" ?

varyvol commented on core/src/main/java/hudson/cli/EnablePluginCommand.java:
```diff
@@ -0,0 +1,80 @@
+        if (restart) {
+            jenkins.safeRestart();
+        }
+        return 0;
```
On other CLI commands I think it's pretty usual to capture errors and return an error code then.

MRamonLeon reviewed (commented):
A single comment: Can we have reciprocal dependencies?

jvz committed:
Fix copyright

Work was done while at work!

jvz committed:
Improve short description of enable-plugin command

jvz committed:
Improve output of enable-plugin command

varyvol commented on core/src/main/java/hudson/cli/EnablePluginCommand.java:
```diff
@@ -0,0 +1,87 @@
+        Jenkins jenkins = Jenkins.get();
+        jenkins.checkPermission(Jenkins.ADMINISTER);
+        PluginManager manager = jenkins.getPluginManager();
+        List<PluginWrapper> pluginsToEnable = pluginNames.stream()
```
Better a set? Different plugins might have the same plugin as a dependency.

jvz replied:
`PluginWrapper` does not implement `equals` or `hashCode`, so it wouldn't really do anything, though I could add them.

varyvol commented on core/src/main/java/hudson/cli/EnablePluginCommand.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static Stream<PluginWrapper> getPluginsToEnable(PluginManager manager, String shortName) {
+        PluginWrapper plugin = manager.getPlugin(shortName);
+        if (plugin == null) throw new IllegalArgumentException(Messages.EnablePluginCommand_NoSuchPlugin(shortName));
```
I'm not sure if it would be useful to distinguish here between a plugin and dependencies? It can be confusing if you request a plugin to be enabled and the error messages talks about other plugin.

jvz committed:
Clarify enable-plugin errors regarding dependencies

jvz committed:
Merge branch 'master' into install-plugin-cli

oleg-nenashev commented:
Will merge tomorrow if there is no negative feedback
